### PR TITLE
Add logic to auto focus HTML editor - DE44817

### PR DESCRIPTION
**Pull request by imcdonald-d2l**
To focus feedback cells in full-size grading view you have to tab into html-editor and then hit enter. This fix auto-focuses the html editor by taking advantage of the newly added `toggleEditing()` function. (See [readme](https://github.com/BrightspaceUI/htmleditor/blob/main/README.md) for details)

[Rally ticket](https://rally1.rallydev.com/#/57732444928d/custom/373260458992?detail=%2Fdefect%2F606965843651)

### Add new feedback
Before | After 
--- | --- 
feedback-not-auto-focused | feedback-auto-focused

### Focus existing feedback 
Before | After 
--- | --- 
feedback-not-auto-focused-2 | feedback-auto-focused-2

### CODE REVIEW

File: d2l-rubric-feedback.js [hunk 1]
```diff
@@ -400,6 +405,13 @@ Polymer({
 		}
 	},
 
+	_focusFeedbackHtmlFromAddFeedbackBtn: function(focus) {
+		if (focus) {
+			setTimeout(() => this._focusHtmlEditor(), 100);
```
Comment (imcdonald-d2l) [inline d2l-rubric-feedback.js:410]: Have to add timeout to give the editor some time to initialize before focusing it.

File: d2l-rubric-feedback.js [hunk 2]
```diff
@@ -331,6 +331,10 @@ Polymer({
 		_disabled: {
```
Comment (imcdonald-d2l) [inline d2l-rubric-feedback.js:331]: `d2l-rubric-feedback`  component only exists in  `d2l-rubric-criteria-group` and `d2l-rubric-criteria-group-mobile` so this shouldn't affect other views. Please correct me if I'm wrong though!

File: d2l-rubric-feedback.js [hunk 3]
```diff
@@ -412,6 +424,14 @@ Polymer({
 		const title = this.localize('criterionFeedback');
 		const extra = this._feedbackHeadingExtra;
 		announce(`${title}${extra}`);
+		this._focusHtmlEditor();
```
Comment (imcdonald-d2l) [inline d2l-rubric-feedback.js:427]: This focuses already existing feedback when tabbing focus.

File: d2l-rubric-feedback.js [hunk 4]
```diff
@@ -412,6 +424,14 @@ Polymer({
...
+	},
+
+	_focusHtmlEditor: function() {
+		if (this.useNewBrightspaceEditor) {
+			const htmLEditor = this._getEditor();
+			if (htmLEditor) htmLEditor.toggleInlineEditing(true);
```
Comment (semibran) [inline d2l-rubric-feedback.js:433]:
```suggestion
			const htmlEditor = this._getEditor();
			if (htmlEditor) htmlEditor.toggleInlineEditing(true);
```

File: d2l-rubric-feedback.js (hunk 1 above)
Comment (semibran) [reply d2l-rubric-feedback.js:410]: super minor nit but I think just `setTimeout(this._focusHtmlEditor, 100)` would work here? Unless that runs into binding issues or something

File: d2l-rubric-feedback.js (hunk 2 above)
Comment (semibran) [reply d2l-rubric-feedback.js:331]: It'd just be `criteria-mobile` not `criteria-group-mobile` but yeah. Looks like there's [a stray import in the editor](https://search.d2l.dev/xref/Brightspace/d2l-rubric/editor/d2l-rubric-criterion-editor.js?r=62e9546a#29) that doesn't get used, that should be the only other place it gets loaded in.

File: d2l-rubric-feedback.js (hunk 4 above)
Comment (imcdonald-d2l) [reply d2l-rubric-feedback.js:433]: This isn't changing anything... did you mean to type something different? 🤔

File: d2l-rubric-feedback.js (hunk 1 above)
Comment (imcdonald-d2l) [reply d2l-rubric-feedback.js]: Yeah, we need the callback function here or we lose the reference to `this`.

File: d2l-rubric-feedback.js (hunk 2 above)
Comment (imcdonald-d2l) [reply d2l-rubric-feedback.js:331]: Yes, my bad. You are correct. Removed the stray import 👍🏻

File: d2l-rubric-feedback.js (hunk 4 above)
Comment (imcdonald-d2l) [reply d2l-rubric-feedback.js:429]: lol, nevermind. I see, I capitalized the `L`

File: d2l-rubric-criteria-mobile.js [hunk 5]
```diff
@@ -176,6 +181,7 @@ Polymer({
 		this._addingFeedback = criterionNum;
 		fastdom.mutate(function() {
 			dom(this.root).querySelector('#feedback' + criterionNum).focus();
+			this._focusFeedbackHtmlEditor = true;
```
Comment (imcdonald-d2l) [inline d2l-rubric-criteria-mobile.js:184]: This adds the focus logic to the "Add Feedback" button in the assignments in-line grading view.

File: editor/d2l-rubric-html-editor.js [hunk 6]
```diff
@@ -160,7 +160,7 @@ Polymer({
 
 	_focusHandler: function() {
 		this._editorActive = true;
-		this._getEditor().focus();
+		this._getEditor().toggleInlineEditing(true);
```
Comment (imcdonald-d2l) [inline editor/d2l-rubric-html-editor.js:163]: The auto-focuses the html-editor for all other instances of the editor. (create rubric experience for example)

File: d2l-rubric-criteria-mobile.js [hunk 7]
```diff
@@ -174,9 +179,7 @@ Polymer({
 	_handleAddFeedback: function(event) {
 		var criterionNum = event.model.get('criterionNum');
 		this._addingFeedback = criterionNum;
-		fastdom.mutate(function() {
-			dom(this.root).querySelector('#feedback' + criterionNum).focus();
-		}.bind(this));
+		this._focusFeedbackHtmlEditor = true;
```
Comment (imcdonald-d2l) [inline d2l-rubric-criteria-mobile.js:182]: This adds the focus logic to the "Add Feedback" button in the assignments in-line grading view.

**Review by semibran — DISMISSED** [review]

**Comment by imcdonald-d2l** [comment]
Not working as expected in in-line grading experience so doing some investigation. 🔍

**Comment by imcdonald-d2l** [comment]
> Not working as expected in in-line grading experience so doing some investigation. 🔍

Issue was that I wasn't setting `notify: true` on property and needed an increase in timeout to successfully focus on editor when clicking on "Add Feedback"

**Review by semibran — DISMISSED** [review]
forcing a delay still feels messy (more so with the magic time delay), is this really the only option?

**Comment by MitchelPaulin** [comment]
Is the large change to the package.json intentional?

**Comment by imcdonald-d2l** [comment]
> Is the large change to the package.json intentional?

Not entirely. Running `npm ci` wasn't getting the html editor update so I had to nuke `package-lock` and re-run `npm i`.

**Comment by imcdonald-d2l** [comment]
> forcing a delay still feels messy (more so with the magic time delay), is this really the only option?

Agreed this is a messy approach. I'll investigate a bit more this morning to see if there is an alternative to the timeout.

**Comment by imcdonald-d2l** [comment]
So after further investigation I noticed that html editor component has a `initializationComplete` promise property which seemed promising at first (pun intended):

image

However, when trying to implement it as follows the promise was not resolving when I hoped it would. It would only resolve after tabbing after hitting the "Add Feedback" button and the html editor was focused manually. Any other thoughts on other approaches instead of using the existing timeout?

```
_focusFeedbackHtmlFromAddFeedbackBtn: async function(focus) {
	if (focus) {
                 // need timeout of 1 to let component render first before getting it
		setTimeout(async () => {
			const htmlEditor = this._getEditor();
			if (htmlEditor && this.useNewBrightspaceEditor) {
				await htmlEditor.initializationComplete;
				this._focusEditor();
			}
		}, 1);
		this.focusFeedbackHtmlEditor = false;
	}
},
```

**Review by devpow112 — APPROVED** [review]

**Comment by semibran** [comment]
> Any other thoughts on other approaches instead of using the existing timeout?

Wasn't able to comment on this before it got merged but just to be anal yeah I'd just use a rAF here or try seeing if a zero timeout still works. Not worth filing a separate PR for obviously but if we ever need to do some more html editor tweaks it'd be a nice-to-have just imo. This new setup is still infinitely better than the old magic timeout we had though 😄

**Pull request merged**